File: 240118/강력한 폭발/strong-explosion.java
import java.io.*;
import java.util.*;

public class Main {

    static int N,K,ans;

    static int[] bombchoice;
    static int[][] map;

    static List<int[]> bomblist;

    static int[][][] pos = {{{-1,0},{-2,0},{1,0},{2,0}},{{-1,0},{0,1},{1,0},{0,-1}},{{-1,-1},{-1,1},{1,-1},{1,1}}};

    static void bomb() {
        boolean[][] check = new boolean[N][N];
        int cnt=0;

        for(int i=0;i<K;i++) {
            int[] tmp = bomblist.get(i);

            int ti = tmp[0];
            int tj = tmp[1];
            int bnum = bombchoice[i];

            for(int d=0;d<4;d++) {
                int ni = ti+pos[bnum][d][0];
                int nj = tj+pos[bnum][d][1];

                if(ni<0 || ni>=N || nj<0 || nj>=N) continue;
                if(map[ni][nj]==1 || check[ni][nj]) continue;

                check[ni][nj] = true;
                cnt++;
            }
        }

        ans = Math.max(cnt,ans);
    }

    static void comb(int cnt, int idx) {
        if(cnt==K) {
            bomb();
            return;
        }

        for(int i=idx;i<3;i++) {
            bombchoice[cnt] = i;
            comb(cnt+1, idx);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());

        map = new int[N][N];
        bomblist = new ArrayList<>();

        StringTokenizer st;
        for(int i=0;i<N;i++) {
            st = new StringTokenizer(br.readLine());
            for(int j=0;j<N;j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
                if(map[i][j]==1) {
                    bomblist.add(new int[] {i,j});
                }
            }
        } //end input

        K = bomblist.size();
        bombchoice = new int[K];
        ans = 0;

        comb(0,0);
        System.out.println(ans+K);
    }
}

File: 240129/강력한 폭발/strong-explosion.py
def bomb():
    global ans
    visit = [[False for _ in range(n)] for _ in range(n)]
    cnt=m

    for i in range(m):
        ti,tj = bomblist[i]
        bnum = check[i]
        
        for d in range(4):
            ni = ti+pos[bnum][d][0]
            nj = tj+pos[bnum][d][1]
            
            if ni<0 or ni>=n or nj<0 or nj>=n:
                continue
            if visit[ni][nj] or board[ni][nj]>0:
                continue
            
            visit[ni][nj] = True
            cnt+=1

    if ans<cnt:
        ans=cnt
        
    return
    

def comb(cnt, idx):
    if cnt==m:
        bomb()
        return
    
    for i in range(idx,3):
        check[cnt] = i
        comb(cnt+1, idx)


n = int(input())
board = [list(map(int, input().split())) for _ in range(n)]
pos = [[(-1,0), (-2,0), (1,0), (2,0)],[(-1,0), (0,-1), (1,0), (0,1)],[(-1,-1), (-1,1), (1,-1), (1,1)]]
ans = 0

bomblist = list()
for i in range(n):
    for j in range(n):
        if board[i][j]==1:
            bomblist.append((i,j))

m = len(bomblist)
check = [0 for _ in range(m)]


comb(0,0)

print(ans)